### PR TITLE
Rules: Size information is returned when listing replication rules #5978

### PR DESCRIPTION
**Pull request by Kwaizer**
Modified `core` to accept the `size` info and return it while listing replication rules

### CODE REVIEW

File: lib/rucio/core/rule.py [hunk 1]
```diff
@@ -28,6 +28,7 @@
 from sqlalchemy import select, update
 from sqlalchemy.exc import IntegrityError, StatementError
 from sqlalchemy.orm.exc import NoResultFound
+from sqlalchemy.orm import joinedload
```
Comment (dchristidis) [inline lib/rucio/core/rule.py]: Unused.

File: lib/rucio/core/rule.py [hunk 2]
```diff
@@ -714,7 +715,10 @@ def list_rules(filters={}, *, session: "Session"):
     :raises:        RucioException
     """
 
-    query = session.query(models.ReplicationRule)
+    query = session.query(models.ReplicationRule).join(models.DataIdentifier,
+                                                       (models.ReplicationRule.scope == models.DataIdentifier.scope) &
+                                                       (models.ReplicationRule.name == models.DataIdentifier.name)
+                                                       ).with_entities(models.ReplicationRule, models.DataIdentifier.bytes)
```
Comment (dchristidis) [inline lib/rucio/core/rule.py]: 1. Is there a reason for `with_entities()`? Would it not work if you added `models.DataIdentifier.bytes` as a second argument to `query()`?
 2. Please prefer the `and_()` function to the `&` operator.
 3. Since you’re touching this area, would you like to try to first migrate it to the SQLAlchemy 2.0 syntax (as its own commit, referencing #5247)? I believe that it will prove trivial.

File: lib/rucio/core/rule.py [hunk 3]
```diff
@@ -750,10 +754,11 @@ def list_rules(filters={}, *, session: "Session"):
             query = query.filter(getattr(models.ReplicationRule, key) == value)
 
     try:
-        for rule in query.yield_per(5):
+        for rule, data_identifier_bytes in query.yield_per(5):
             d = {}
             for column in rule.__table__.columns:
                 d[column.name] = getattr(rule, column.name)
```
Comment (dchristidis) [inline lib/rucio/core/rule.py]: Please take the opportunity to replace with `to_dict()` (in a separate commit).

File: bin/rucio [hunk 4]
```diff
@@ -1490,9 +1491,10 @@ def list_rules(args):
                           '%s[%d/%d/%d]' % (rule['state'], rule['locks_ok_cnt'], rule['locks_replicating_cnt'], rule['locks_stuck_cnt']),
                           rule['rse_expression'],
                           rule['copies'],
+                          rule['size'],
```
Comment (dchristidis) [inline bin/rucio]: I am concerned about the output of `list-rules` being already very long in terms of width. Also, understanding the magnitude from bytes takes effort. I think it will be preferable to replace this with `sizefmt(rule['size']) if rule['size'] is not None else 'N/A'` (and adapt the header accordingly).

File: lib/rucio/core/rule.py [hunk 5]
```diff
@@ -750,10 +754,11 @@ def list_rules(filters={}, *, session: "Session"):
...
+            d['size'] = data_identifier_bytes
```
Comment (dchristidis) [inline lib/rucio/core/rule.py]: I wonder whether `bytes` is better than `size` because it’s more consistent with all other attributes that represent volume.

**Review by dchristidis — CHANGES_REQUESTED** [review]

File: lib/rucio/core/rule.py (hunk 2 above)
Comment (Kwaizer) [reply lib/rucio/core/rule.py]: 1. Initially I was considering using the second argument but it evoked several `TypeError`s in tests. But I'll try to change it with two arguments again
3 .So it should be redone like: 
```
query = select(models.ReplicationRule, models.DataIdentifier.bytes).join(
models.DataIdentifier, and_(
(models.ReplicationRule.scope == models.DataIdentifier.scope),
(models.ReplicationRule.name == models.DataIdentifier.name))
```

right?

File: lib/rucio/core/rule.py (hunk 3 above)
Comment (Kwaizer) [reply lib/rucio/core/rule.py]: So I need to make another commit to this pr and then just squash all the commits?

File: lib/rucio/core/rule.py (hunk 2 above)
Comment (dchristidis) [reply lib/rucio/core/rule.py]: Yes, like so. You can remove the parentheses around the comparisons. I hope I’m not misleading you, but a quick attempt of my own appeared to be successful.

File: lib/rucio/core/rule.py (hunk 3 above)
Comment (dchristidis) [reply lib/rucio/core/rule.py]: I always recommend changes that are supposed to maintain exact functionality to be isolated and merged as separate commits (and not squashed). This aids the review process, but also future debugging. Consequently, I find it reasonable for this PR to be made out of three commits: the SQLAlchemy 2.0 migration, this replacement with `to_dict()`, and finally the changes that close the issue related to this PR.

File: lib/rucio/core/rule.py (hunk 2 above)
Comment (dchristidis) [reply lib/rucio/core/rule.py]: Please also replace all instances of the `filter()` method with `where()`.

We’ve adopted a different styling convention for SQL queries which improves their readability. Consequently, I would recommend formatting it like so:
```python
stmt = select(
    models.ReplicationRule,
    models.DataIdentifier.bytes
).join(
    models.DataIdentifier,
    and_(
        models.ReplicationRule.scope == models.DataIdentifier.scope,
        models.ReplicationRule.name == models.DataIdentifier.name
    )
)
```

File: lib/rucio/core/rule.py (hunk 3 above)
Comment (dchristidis) [reply lib/rucio/core/rule.py]: You’ll need to take some time to rewrite your commit history in order to achieve what I described above.

Comment (Kwaizer) [reply lib/rucio/core/rule.py]: Sure, no problem

Comment (dchristidis) [reply lib/rucio/core/rule.py]: Despite the latest changes, the commit history has room for improvement.

 1. The subject of the first commit is `Rules: SQLAlchemy 2.0 migration #5978`, yet it modifies `bin/rucio` and changes the structure return by `list_rules()`. Also, it introduces some code formatted unconventionally, that is then improved by the third commit.
 2. The second commit is perfect.
 3. The third commit will have to be adjusted, once the first commit is fixed.

In short, the end result (meaning the three commits combined) is perfect. However, your individual commits are introducing some mistakes (e.g. `size` instead of `bytes`) and fixing them afterwards whereas you should be introducing the final iteration of your code immediately.

Let me know in case I’m not being clear or you require assistance.

Comment (Kwaizer) [reply lib/rucio/core/rule.py]: I rewrote commit history. Please, have a look, it seems to be complete: 

I deleted and edited changes in my local branch to the point it had only SQLAlchemy 2.0 migration, pushed it, added `to_dict()` - pushed another commit 
and then made final commit. 

In total I had 6 commits in PR and squashed 3 of them. Such operations don't seem to be very effective. Could I have changed commit history easier and faster?

Comment (dchristidis) [reply lib/rucio/core/rule.py]: Yes, this is exactly what I was hoping for. Thanks!

`git rebase --interactive` allows all sort of rewrites of your commit history (e.g. edit, merge, split, reorder, and drop commits) in a systematic way. In this particular case, I would have expected something along the lines of:

 1. Start the rebase at the base of your branch, `edit` the first and third commit and `pick` the second commit.
 2. Edit the two files as appropriately, then amend the commit. Alternatively, undo the commit with `git reset`, edit, then commit anew.
 3. Continue the rebase. The second commit should be applied automatically.
 4. The third commit will pose a conflict, due to the changes applied on the first commit. Simply edit (and reintroduce the changes to `bin/rucio` which were removed in the first commit) until you’re happy with the end result.
 5. Finish the rebase.

It might be worth looking up for some specific examples first, but once gotten used to, it will become second nature.

Feel free to mark this conversation as resolved after you’re read my response.

**Review by dchristidis — APPROVED** [review]

File: lib/rucio/core/rule.py (hunk 3 above)
Comment (Kwaizer) [reply lib/rucio/core/rule.py]: Such approach seems to be more convenient. Thanks! I'll keep that in mind.

File: bin/rucio [hunk 6]
```diff
@@ -1478,6 +1478,7 @@ def list_rules(args):
                   '%s[%d/%d/%d]' % (rule['state'], rule['locks_ok_cnt'], rule['locks_replicating_cnt'], rule['locks_stuck_cnt']),
                   rule['rse_expression'],
                   rule['copies'],
+                  sizefmt(rule['bytes']) if rule['bytes'] is not None else 'N/A',
```
Comment (dchristidis) [inline bin/rucio]: Forgive me, I forgot about something.

https://github.com/rucio/rucio/blob/a73004e1e6ecc5815c2c11a4c80e641be0b97930/bin/rucio#L1861

For completeness, you need to pass `args.human` as the second argument to `sizefmt()`. This also applies to the other call too.

**Review by dchristidis — CHANGES_REQUESTED** [review]

**Review by dchristidis — APPROVED** [review]
